nntrainer/optimizers/adamw.cpp

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2024 Daniel Jang <minhyukjang@snu.ac.kr>
+ *
+ * @file   adamw.cpp
+ * @date   3 November 2024
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jijoong Moon <jijoong.moon@samsung.com>
+ * @author Parichay Kapoor <pk.kapoor@samsung.com>
+ * @author Daniel Jang <minhyukjang@snu.ac.kr>
+ * @bug    No known bugs except for NYI items
+ * @brief  This is the AdamW Optimizer.
+ */
+
+#include <cmath>
+#include <fstream>
+
+#include <adamw.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <node_exporter.h>
+#include <util_func.h>
+
+namespace nntrainer {
+
+AdamW::AdamW() : adam_props(PropsB1(), PropsB2(), PropsEpsilon(), TorchRef()) {
```

Comment:
Adam case, when the torch_ref flag is provided, it handles separate operations to ensure that the results are as similar as possible to PyTorch.

I am curious whether Adamw also considered this case in adding torch_ref or simply did so to have the same properties as Adam. It seems like torch_ref has no meaning for AdamW.

```.cpp
  if (torch_ref) {
    Tensor denom = wv.apply<float>(sqrtFloat<float>);
    denom.divide_i(sqrtFloat(biasCorrection2));
    denom.add_i(epsilon);
    wm.divide(denom, x_grad);

    context.applyGradient(context.getLearningRate() / biasCorrection1);

  } else {
    std::function<double(double)> sqrtEps = [epsilon](double f) {
      return 1 / (sqrtDouble(f) + epsilon);
    };
```

Reply:
Indeed, the `"torch_ref"` property is not used (at all) in `AdamW`. The reasoning behind keeping it (and also retaining the identifier `adam_props`, to address @baek2sm's [comment](https://github.com/nnstreamer/nntrainer/pull/2780#discussion_r1825411607)) was that I thought it would be a very good idea to, in a later PR, have the `AdamW` class inherit from the `Adam` class, to reduce code duplication, as they share a lot of boilerplate code. Sorry for not making this clear.

Any thoughts on this? If this is not the way to go, I'd happily make the necessary changes.

FYI: PyTorch's [AdamW class](https://pytorch.org/docs/stable/_modules/torch/optim/adamw.html#AdamW) <ins>doesn't</ins> seem to inherit from its [Adam class](https://pytorch.org/docs/stable/_modules/torch/optim/adam.html#Adam).